kani-driver/src/concrete_playback.rs

```diff
@@ -256,156 +281,127 @@ struct UnitTest {
-
-    /// The first-level CBMC output parser. This extracts the result message.
-    fn parse_cbmc_out(cbmc_out: &Value) -> Result<Vec<ConcreteVal>> {
+    /// The first-level extractor. Traverses processed items to find properties.
+    pub fn extract_from_processed_items(
+        processed_items: &[ParserItem],
+    ) -> Result<Vec<ConcreteVal>> {
         let mut concrete_vals: Vec<ConcreteVal> = Vec::new();
-        let cbmc_out_arr = cbmc_out
-            .as_array()
-            .with_context(|| format!("Expected this CBMC output to be an array: {}", cbmc_out))?;
-        let mut have_parsed_assert_fail = false;
-        for general_msg in cbmc_out_arr {
-            let result_msg = &general_msg["result"];
-            if !result_msg.is_null() {
-                let result_arr = result_msg.as_array().with_context(|| {
-                    format!("Expected this CBMC result object to be an array: {}", result_msg)
-                })?;
-                for result_val in result_arr {
-                    parse_result(result_val, &mut concrete_vals, &mut have_parsed_assert_fail)?;
+        let mut extracted_assert_fail = false;
+        for processed_item in processed_items {
+            if let ParserItem::Result { result } = processed_item {
+                for property in result {
+                    // Even after extracting an assert fail, we continue to call extract on more properties to provide
+                    // better diagnostics to the user in case they expected even future checks to be extracted.
+                    let old_extracted_assert_fail = extracted_assert_fail;
+                    let new_concrete_vals =
+                        extract_from_property(property, &mut extracted_assert_fail)?;
+                    if !old_extracted_assert_fail && extracted_assert_fail {
+                        concrete_vals = new_concrete_vals;
+                    }
                 }
             }
         }
         Ok(concrete_vals)
     }
 
-    /// The second-level CBMC output parser. This extracts the trace entries of failing assertions.
-    fn parse_result(
-        result_val: &Value,
-        concrete_vals: &mut Vec<ConcreteVal>,
-        have_parsed_assert_fail: &mut bool,
-    ) -> Result<()> {
-        let desc = result_val["description"].to_string();
-        let prop = result_val["property"].to_string();
-        let status = result_val["status"].to_string();
-        let prop_is_assert = prop.contains("assertion");
-        let status_is_failure = status == "\"FAILURE\"";
-
-        if prop_is_assert && status_is_failure {
-            if *have_parsed_assert_fail {
+    /// The second-level extractor. Traverses properties to find trace items.
+    fn extract_from_property(
+        property: &Property,
+        extracted_assert_fail: &mut bool,
+    ) -> Result<Vec<ConcreteVal>> {
+        let mut concrete_vals: Vec<ConcreteVal> = Vec::new();
+        let property_is_assert = property.property.contains("assertion");
```

Comment:
Wouldn't it be better to check the property class (part of `property`)?

Reply:
Do you mean the `extract_property_class` function? I updated the code to use that function.